### DISCUSSION TITLE
Workflow question: acting on bibtex entries

### DISCUSSION THREAD

**Discussion by EFLS** [emacs-citar/citar#564] [Q&A] [answered]
Hi

This is a simple question about workflow. I've integrated `citar` into my citing and note taking routine, working with `org-cite` elements is great (opening PDFs etc.). Works great!

What I'm wondering: is there a way to do some of the citar actions from a bibtex file? I manage my bibfiles in Emacs, so it would be really handy if I were able to open, for example, attached files from the bibtex file itself. Is there any built-in way to do this? Or how do others deal with these use-cases?

Cheers
EFLS

**Comment by bdarcus** [comment, marked as answer]
https://github.com/bdarcus/citar/wiki/Citar-in-bibtex-files-with-Embark

> if I were able to open, for example, attached files from the bibtex file itself. Is there any built-in way to do this? Or how do others deal with these use-cases?

No, but conceptually it's pretty simple.

There are two approaches.

# Embark

Here you'd want an embark target finder for BibTeX keys, and then to associate a keymap with that type, with whatever commands you want available via `embark-act`.

Citar includes similar target finders for markdown, latex, and org citation keys.

Are you comfortable trying to code that? If yes, this would be a good example to add to the wiki.

Here's a start, adapted from [here](https://github.com/tmalsburg/helm-bibtex/issues/121#issuecomment-237981605):

``` elisp
(defun bibtex-key-at-point ()
  (save-excursion
    (bibtex-beginning-of-entry)
    (when (looking-at bibtex-entry-maybe-empty-head)
      (bibtex-key-in-head))))

(add-to-list 'embark-target-finders 'bibtex-key-at-point)
(add-to-list 'embark-keymap-alist '(bibtex-key . bibtex-map))
```
# A dedicated command

The other approach is a dedicated interactive command, like the one I linked to above.

**Reply by EFLS** [reply]
This looks great, thanks. I'll check this out and report back.

**Reply by bdarcus** [reply]
Note that as I look at it again, I think there's one obvious piece missing on the embark example: the function isn't typing the found key as a "bibtex-key". Hopefully you can figure that out.

**Reply by EFLS** [reply]
I've got this:

``` elisp
(defun bibtex-key-embark ()  
  (save-excursion  
    (bibtex-beginning-of-entry)  
    (when (looking-at bibtex-entry-maybe-empty-head)  
      (cons 'bibtex-key
            (bibtex-key-in-head)))))

(with-eval-after-load "embark"
  (embark-define-keymap bibtex-key-embark-map
    "Embark keymap for Zetteldeft links")
  (general-define-key :keymaps 'bibtex-key-embark-map
                      "f" 'citar-open
                      "n" 'citar-open-notes))

(with-eval-after-load "embark"
  (add-to-list 'embark-target-finders 'bibtex-key-embark)  
  (add-to-list 'embark-keymap-alist '(bibtex-key . bibtex-key-embark-map)))
```


I use `general-define-key` because I forgot how to bind keys without 🙃

**Reply by bdarcus** [reply]
So it works as expected for you?

**Reply by EFLS** [reply]
Yeah it does. Is there a wiki already where I could add it? Perhaps there can be other functions added to the Embark map

**Reply by bdarcus** [reply]
Yep:

https://github.com/bdarcus/citar/wiki/Embark

Can you see if you can edit that by adding this? I'm a little uncertain of the permissions with the github wikis.

**Reply by EFLS** [reply]
I've included some explanation and fixed the keybindings. Added it to a separate page (as I feel it's different from the Embark stuff), but obviously feel free to move it to where you see fit.
